ios/CRLChecker.mm

```diff
@@ -0,0 +1,173 @@
+    }
+
+    EVP_PKEY_free(pubkey);
+    X509_free(issuer);
+  }
+
+  // --- Check CRL validity period ---
+  const ASN1_TIME *lastUpdate = X509_CRL_get0_lastUpdate(crl);
+  const ASN1_TIME *nextUpdate = X509_CRL_get0_nextUpdate(crl);
+  if (!lastUpdate || !nextUpdate) {
+    X509_free(cert);
+    X509_CRL_free(crl);
+    return -4;
+  }
+
+  if (X509_cmp_current_time(lastUpdate) > 0 ||
+      X509_cmp_current_time(nextUpdate) < 0) {
+    X509_free(cert);
+    X509_CRL_free(crl);
+    return -4; // CRL is not currently valid
+  }
+
+  // --- Check CRL issuer matches cert issuer ---
+  if (issuer_der_or_null != NULL && issuer_len > 0) {
+    const X509_NAME *crlIssuer = X509_CRL_get_issuer(crl);
+    const X509_NAME *certIssuer = X509_get_issuer_name(cert);
+    if (X509_NAME_cmp(crlIssuer, certIssuer) != 0) {
+      X509_free(cert);
+      X509_CRL_free(crl);
+      return -5; // CRL issuer mismatch
+    }
+  }
+
+  // --- Check if the certificate is listed in the CRL ---
+  ASN1_INTEGER *serial = X509_get_serialNumber(cert);
+  STACK_OF(X509_REVOKED) *revoked = X509_CRL_get_REVOKED(crl);
+
+  BOOL isRevoked = NO;
+  for (int i = 0; i < sk_X509_REVOKED_num(revoked); i++) {
+    X509_REVOKED *rev = sk_X509_REVOKED_value(revoked, i);
+    if (ASN1_INTEGER_cmp(serial, X509_REVOKED_get0_serialNumber(rev)) == 0) {
+      isRevoked = YES;
+      break;
+    }
+  }
+
+  X509_free(cert);
+  X509_CRL_free(crl);
+  return isRevoked ? 1 : 0;
+}
+
+/**
+ * Extracts the first available CRL Distribution Point URI from a DER-encoded
+ * X.509 certificate. This URI is where the CRL (Certificate Revocation List)
+ * can be downloaded from.
+ *
+ * @param cert_der Pointer to the DER-encoded certificate.
+ * @param cert_len Length of the certificate in bytes.
+ *
+ * @return const char* A malloc'd null-terminated string containing the CRL
+ * Distribution Point URL, or NULL if none is found. The caller is responsible
+ * for freeing the memory.
+ */
+const char *extractCRLFromCert(const unsigned char *cert_der, int cert_len) {
+  const unsigned char *ptr = cert_der;
+  X509 *cert = d2i_X509(NULL, &ptr, cert_len);
+  if (!cert)
+    return NULL;
+
+  STACK_OF(DIST_POINT) *crl_dp = static_cast<STACK_OF(DIST_POINT) *>(
+      X509_get_ext_d2i(cert, NID_crl_distribution_points, NULL, NULL));
+
+  if (!crl_dp || sk_DIST_POINT_num(crl_dp) == 0) {
+    X509_free(cert);
+    return NULL;
+  }
+
+  for (int i = 0; i < sk_DIST_POINT_num(crl_dp); i++) {
+    DIST_POINT *dp = sk_DIST_POINT_value(crl_dp, i);
+    if (dp->distpoint && dp->distpoint->type == 0 &&
+        dp->distpoint->name.fullname) {
+      GENERAL_NAMES *names = dp->distpoint->name.fullname;
+      for (int j = 0; j < sk_GENERAL_NAME_num(names); j++) {
+        GENERAL_NAME *gen = sk_GENERAL_NAME_value(names, j);
+        if (gen->type == GEN_URI) {
+          ASN1_IA5STRING *uri = gen->d.uniformResourceIdentifier;
+          char *url = (char *)malloc(uri->length + 1);
```

Comment:
This is the `malloc` referenced in the documentation comment. Keep this in mind.